### PR TITLE
Update kafka event listener to new schema

### LINKED ISSUE #19358: Update KafkaEventListener to use new cross-institution message schema
Use new event schema from here: https://github.com/ess-dmsc/streaming-data-types
Seeking to start of run should be done by timestamp in run messages rather than offset.

These changes will permit multipartition streaming for event data and make the listener suitable for use by ESS and SINQ as well as ISIS.

### PR DESCRIPTION
**Pull request by matthew-d-jones**
Updated the `KafkaEventListener` to use new schema for neutron event data which is designed for use by PSI, ESS and ISIS.
As part of this change the event data topic can consist of multiple partitions, allowing streaming through multiple brokers in a Kafka cluster to achieve higher throughput.

Removed mention of the Kafka listener from the release notes as it is still undergoing major changes, including to the interface. This was agreed on at the IBEX-Kafka-Mantid meeting 3 Mar 2017.

**To test:**

*To test this PR you must be at ISIS*

Launch MantidPlot and select "TEST_LIVE" as your facility and "ISIS_Kafka_Event" as your instrument. From the "Workspaces" view (crtl+shift+m) click "Load" -> "Live Data". Match the settings in the following screenshot and then hit run, after around 30 seconds the stream should end, check you have a workspace with 14258850 events in it.

livedatatest

Fixes #19358.

*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?

- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by martyngigg** [comment]
@matthew-d-jones If this is marked for 3.11 then the release note change won't get in. Was that intended?

**Comment by matthew-d-jones** [comment]
@martyngigg That wasn't intended, I forgot about the release note change. It would be best to get this into 3.10.

**Comment by martyngigg** [comment]
@matthew-d-jones Thanks for the clarification. I'll get on and review it.

File: Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp [hunk 1]
```diff
@@ -348,10 +252,141 @@ void KafkaEventStreamDecoder::captureImplExcept() {
...
+      auto &periodBuffer =
+          *m_localEvents[static_cast<size_t>(ISISMsg->period_number())];
+      auto &mutableRunInfo = periodBuffer.mutableRun();
+      mutableRunInfo.getTimeSeriesProperty<double>(PROTON_CHARGE_PROPERTY)
+          ->addValue(pulseTime, ISISMsg->proton_charge());
+      for (decltype(nEvents) i = 0; i < nEvents; ++i) {
+        auto &spectrum = periodBuffer.getSpectrum(
+            m_specToIdx[static_cast<int32_t>(detData[i])]);
+        spectrum.addEventQuickly(TofEvent(static_cast<double>(tofData[i]) *
+                                              1e-9, // nanoseconds to seconds
```
Comment (matthew-d-jones) [inline Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp]: This should be `1e-3, // nanoseconds to microseconds`

**Review by martyngigg — CHANGES_REQUESTED** [review]
Code generally looks good.

Just a few minor style points suggested. I'll do some functional testing now.

**Comment by martyngigg** [comment]
I'm just running through some tests and I don't quite see the expected behaviour.

After a while I get a workspace with only `9496266` events. Also when I try to cancel `MonitorLiveData` it gets stuck and won't stop. `gdb` indicates that it's stuck in `rdkafka`:

```
(gdb) thread 12
[Switching to thread 12 (Thread 0x7f4d337fe700 (LWP 6386))]
#0  0x00007f4d9ffaf98d in pthread_join (threadid=139969393833728, thread_return=0x7f4d337fd200) at pthread_join.c:90
90      pthread_join.c: No such file or directory.
(gdb) bt
#0  0x00007f4d9ffaf98d in pthread_join (threadid=139969393833728, thread_return=0x7f4d337fd200) at pthread_join.c:90
#1  0x00007f4d6ca3a700 in ?? () from /usr/lib/x86_64-linux-gnu/librdkafka.so.1
#2  0x00007f4d6c9e5f94 in rd_kafka_destroy () from /usr/lib/x86_64-linux-gnu/librdkafka.so.1
#3  0x00007f4d6cc8259e in RdKafka::KafkaConsumerImpl::close() () from /usr/lib/x86_64-linux-gnu/librdkafka++.so.1
#4  0x00007f4d6cfca560 in Mantid::LiveData::KafkaTopicSubscriber::~KafkaTopicSubscriber (this=0x7f4d680057c0, __in_chrg=<optimised out>) at ../../Framework/LiveData/src/Kafka/KafkaTopicSubscriber.cpp:81
#5  0x00007f4d6cfca5da in Mantid::LiveData::KafkaTopicSubscriber::~KafkaTopicSubscriber (this=0x7f4d680057c0, __in_chrg=<optimised out>) at ../../Framework/LiveData/src/Kafka/KafkaTopicSubscriber.cpp:91
#6  0x00007f4d6cfc0744 in std::default_delete<Mantid::LiveData::IKafkaStreamSubscriber>::operator() (this=0x7f4d68002a88, __ptr=0x7f4d680057c0) at /usr/include/c++/5/bits/unique_ptr.h:76
#7  0x00007f4d6cfbee5d in std::unique_ptr<Mantid::LiveData::IKafkaStreamSubscriber, std::default_delete<Mantid::LiveData::IKafkaStreamSubscriber> >::~unique_ptr (this=0x7f4d68002a88, __in_chrg=<optimised out>)
    at /usr/include/c++/5/bits/unique_ptr.h:236
#8  0x00007f4d6cfb7fb8 in Mantid::LiveData::KafkaEventStreamDecoder::~KafkaEventStreamDecoder (this=0x7f4d68002a10, __in_chrg=<optimised out>)
    at ../../Framework/LiveData/src/Kafka/KafkaEventStreamDecoder.cpp:71
#9  0x00007f4d6cfb6616 in std::default_delete<Mantid::LiveData::KafkaEventStreamDecoder>::operator() (this=0x7f4d68004128, __ptr=0x7f4d68002a10) at /usr/include/c++/5/bits/unique_ptr.h:76
#10 0x00007f4d6cfb5da1 in std::unique_ptr<Mantid::LiveData::KafkaEventStreamDecoder, std::default_delete<Mantid::LiveData::KafkaEventStreamDecoder> >::~unique_ptr (this=0x7f4d68004128, 
    __in_chrg=<optimised out>) at /usr/include/c++/5/bits/unique_ptr.h:236
#11 0x00007f4d6cfb74e4 in Mantid::LiveData::KafkaEventListener::~KafkaEventListener (this=0x7f4d680040d0, __in_chrg=<optimised out>, __vtt_parm=<optimised out>)
    at ../../Framework/LiveData/inc/MantidLiveData/Kafka/KafkaEventListener.h:43
#12 0x00007f4d6cfb7554 in Mantid::LiveData::KafkaEventListener::~KafkaEventListener (this=0x7f4d680040d0, __in_chrg=<optimised out>, __vtt_parm=<optimised out>)
    at ../../Framework/LiveData/inc/MantidLiveData/Kafka/KafkaEventListener.h:43
#13 0x00007f4d6cfb7a1f in boost::checked_delete<Mantid::LiveData::KafkaEventListener> (x=0x7f4d680040d0) at /usr/include/boost/core/checked_delete.hpp:34
#14 0x00007f4d6cfb7aac in boost::detail::sp_counted_impl_p<Mantid::LiveData::KafkaEventListener>::dispose (this=0x7f4d68001710) at /usr/include/boost/smart_ptr/detail/sp_counted_impl.hpp:78
#15 0x00007f4d6cf519d2 in boost::detail::sp_counted_base::release (this=0x7f4d68001710) at /usr/include/boost/smart_ptr/detail/sp_counted_base_gcc_x86.hpp:146
#16 0x00007f4d6cf51a65 in boost::detail::shared_count::~shared_count (this=0x7f4d337fd508, __in_chrg=<optimised out>) at /usr/include/boost/smart_ptr/detail/shared_count.hpp:443
#17 0x00007f4d6cf89ad4 in boost::shared_ptr<Mantid::API::ILiveListener>::~shared_ptr (this=0x7f4d337fd500, __in_chrg=<optimised out>) at /usr/include/boost/smart_ptr/shared_ptr.hpp:323
#18 0x00007f4d6cf9da1b in Mantid::LiveData::MonitorLiveData::exec (this=0x7f4d683f4230) at ../../Framework/LiveData/src/MonitorLiveData.cpp:110
#19 0x00007f4da6f584b4 in Mantid::API::Algorithm::execute (this=0x7f4d683f4230) at ../../Framework/API/src/Algorithm.cpp:548
#20 0x00007f4da6f5df8d in Mantid::API::Algorithm::executeAsyncImpl (this=0x7f4d683f4230) at ../../Framework/API/src/Algorithm.cpp:1491
#21 0x00007f4da6f71e82 in Poco::ActiveRunnable<bool, Poco::Void, Mantid::API::Algorithm>::run (this=0x7f4d683f8080) at /usr/include/Poco/ActiveRunnable.h:65
#22 0x00007f4da5352d82 in Poco::PooledThread::run() () from /usr/lib/libPocoFoundationd.so.31
#23 0x00007f4da534fd8d in ?? () from /usr/lib/libPocoFoundationd.so.31
#24 0x00007f4da534fa34 in Poco::ThreadImpl::runnableEntry(void*) () from /usr/lib/libPocoFoundationd.so.31
#25 0x00007f4d9ffae6ba in start_thread (arg=0x7f4d337fe700) at pthread_create.c:333
#26 0x00007f4d9fce482d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
(gdb) thread 13
[Switching to thread 13 (Thread 0x7f4d29ffb700 (LWP 6393))]
#0  0x00007f4d9ffaf98d in pthread_join (threadid=139968815032064, thread_return=0x7f4d29ffac80) at pthread_join.c:90
90      in pthread_join.c
(gdb) bt
#0  0x00007f4d9ffaf98d in pthread_join (threadid=139968815032064, thread_return=0x7f4d29ffac80) at pthread_join.c:90
#1  0x00007f4d6ca3a700 in ?? () from /usr/lib/x86_64-linux-gnu/librdkafka.so.1
#2  0x00007f4d6c9e1705 in ?? () from /usr/lib/x86_64-linux-gnu/librdkafka.so.1
#3  0x00007f4d6c9e2591 in ?? () from /usr/lib/x86_64-linux-gnu/librdkafka.so.1
#4  0x00007f4d6ca3a237 in ?? () from /usr/lib/x86_64-linux-gnu/librdkafka.so.1
#5  0x00007f4d9ffae6ba in start_thread (arg=0x7f4d29ffb700) at pthread_create.c:333
#6  0x00007f4d9fce482d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
(gdb) 
```
Have you observed this?

**Comment by matthew-d-jones** [comment]
The changes in this PR require a bug fix which will be in the next release of `librdkafka` (v0.9.6).

**Review by peterfpeterson — DISMISSED** [review]
This is a test

**Comment by matthew-d-jones** [comment]
@mantid-builder I've got conferences for the next two weeks, after that I'll look into updating `librdkafka` and sorting out this PR.

**Pull request closed without merging**